File: problems/16.graphs/ts/network-delay.ts
import { PriorityQueueImpl } from './utils/PriorityQueueImpl';

export function networkDelay(times: number[][], k: number, n: number) {
    const adjList = Array.from({ length: n }, () => Array<number[]>());
    const delays = Array.from({ length: n }, () => Infinity);

    for (let i = 0; i < times.length; i++) {
        const [S, T, W] = times[i];
        adjList[S - 1].push([T - 1, W]);
    }
    delays[k - 1] = 0;

    const heap = new PriorityQueueImpl([], (a, b) => delays[a] < delays[b]);
    heap.insert(k - 1);

    while (!heap.isEmpty()) {
        const current = heap.pop();
        if (current === undefined) break;
        const connections = adjList[current];

        for (let i = 0; i < connections.length; i++) {
            const [connection, weight] = connections[i];
            if (
                connection !== undefined &&
                weight + delays[current] < delays[connection]
            ) {
                delays[connection] = weight + delays[current];
                heap.insert(connection);
            }
        }
    }

    return Math.max.apply(null, delays) === Infinity
        ? -1
        : Math.max.apply(null, delays);
}

File: problems/16.graphs/golang/main/network_delay.go
package main

import (
	"github.com/AmirAhmadzadeh/golang/graph-algo/priority_queue"
	"math"
)

/// dijkstra
func networkDelay1(times [][]int, k int, n int) int {
	adjList := make([][][]int, n)
	for _, time := range times {
		source := time[0] - 1
		target := time[1] - 1
		weight := time[2]

		adjList[source] = append(adjList[source], []int{target, weight})
	}

	delays := []float64{}
	for i := 0; i < n; i++ {
		delays = append(delays, math.Inf(1))
	}
	delays[k-1] = 0

	pq := priority_queue.NewPriorityQueue([]int{}, func(a, b int) bool {
		return delays[a] < delays[b]
	})
	pq.Insert(k - 1)

	for !pq.IsEmpty() {
		current, error := pq.Peek()
		if error != nil {
			break
		}
		connections := adjList[current]

		for _, connection := range connections {
			T := connection[0]
			W := connection[1]
			if float64(W)+delays[current] < delays[T] {
				delays[T] = float64(W) + delays[current]
				pq.Insert(T)
			}
		}
	}

	return int(Max(delays))
}

func networkDelay2(times [][]int, k int, n int) int {
	delays := []float64{}
	for i := 0; i < n; i++ {
		delays = append(delays, math.Inf(1))
	}
	delays[k-1] = 0

	for i := 0; i < n-1; i++ {
		count := 0
		for _, time := range times {
			source := time[0] - 1
			target := time[1] - 1
			weight := time[2]

			if float64(weight)+delays[source] < delays[target] {
				delays[target] = delays[source] + float64(weight)
				count += 1
			}
		}

		if count == 0 {
			break
		}
	}

	if Max(delays) == math.Inf(1) {
		return -1
	}

	return int(Max(delays))
}

func Max(slice []float64) float64 {
	max := slice[0]
	for _, value := range slice {
		if max < value {
			max = value
		}
	}
	return max
}
